### ISSUE TITLE
SIMD 64 Bit Memory-Alignment Bug 

### ISSUE THREAD

**Issue by Gallimathias** [dotnet/coreclr#15906]
Since I was a little wrong with issue https://github.com/dotnet/roslyn/issues/24233#issuecomment-358405137 in the Roslyn repo, I apologize for that ^^, here at the right place.

**Version Used**: .NET 4.7.1 C# 7.2

**Steps to Reproduce**:

1. Install engenious via nuget or wrote your own packed Vector3 or use a other packed Vector3
2. Cast with pointer to System.Numerics.Vector3 or own unpacked Vector3
3.  Just do the same thing, just put the pointer in a variable before it.
4. Compare the results

**A brief example:**

C#-Code:

````CSharp
        static unsafe void Test()
        {
            var value = new Vector3(0, 0, -41);
            var a = *(Numerics.Vector3*)(&value);
        }

        static unsafe void TestB()
        {
            var value = new Vector3(0, 0, -41);
            var d = &value;
            var a = *(Numerics.Vector3*)d;
        }
````

IL-Code:

````CIL
                 .method private hidebysig static 
			void Test () cil managed 
		{
			// Method begins at RVA 0x2144
			// Code size 33 (0x21)
			.maxstack 4
			.locals init (
				[0] valuetype [engenious]engenious.Vector3 'value',
				[1] valuetype [System.Numerics.Vectors]System.Numerics.Vector3 a
			)

			IL_0000: nop
			IL_0001: ldloca.s 'value'
			IL_0003: ldc.r4 0.0
			IL_0008: ldc.r4 0.0
			IL_000d: ldc.r4 -41
			IL_0012: call instance void [engenious]engenious.Vector3::.ctor(float32, float32, float32)
			IL_0017: ldloca.s 'value'
			IL_0019: conv.u
			IL_001a: ldobj [System.Numerics.Vectors]System.Numerics.Vector3
			IL_001f: stloc.1
			IL_0020: ret
		} // end of method Program::Test

		.method private hidebysig static 
			void TestB () cil managed 
		{
			// Method begins at RVA 0x2174
			// Code size 35 (0x23)
			.maxstack 4
			.locals init (
				[0] valuetype [engenious]engenious.Vector3 'value',
				[1] valuetype [engenious]engenious.Vector3* d,
				[2] valuetype [System.Numerics.Vectors]System.Numerics.Vector3 a
			)

			IL_0000: nop
			IL_0001: ldloca.s 'value'
			IL_0003: ldc.r4 0.0
			IL_0008: ldc.r4 0.0
			IL_000d: ldc.r4 -41
			IL_0012: call instance void [engenious]engenious.Vector3::.ctor(float32, float32, float32)
			IL_0017: ldloca.s 'value'
			IL_0019: conv.u
			IL_001a: stloc.1
			IL_001b: ldloc.1
			IL_001c: ldobj [System.Numerics.Vectors]System.Numerics.Vector3
			IL_0021: stloc.2
			IL_0022: ret
		} // end of method Program::TestB
````

**Expected Behavior**: Both methods should produce the same result on all systems.

Method Test: a should be <0. 0. -41>
Method TestB: a should be <0. 0. -41>

**Actual Behavior**: When you cast a packed vector into an unpacked vector, the result will always be a zero vector on some computers with 64 Bit and the use of a direct assignment.

Method TestB works on all systems no matter if 32 Bit or 64 Bit.
Method Test does not work as 64 bit version on some systems. Could this be related to the CPU used?

My Results compiled as 64 Bit:

Method Test: a is <0. 0. 0>
Metod TestB: a is <0. 0. -41>

On my PC, Method Test doesn't work. My CPU is Intel i7-5960X.

The only difference in CIL are the following lines:

IL_001a: stloc.1
IL_001b: ldloc.1

**Supplementary here the assembly**

__Method TestB:__
````Assembly
 var d = &value;
00007FF958820875  lea         rax,[rbp+38h]  
00007FF958820879  mov         qword ptr [rbp+30h],rax  
            var a = *(Numerics.Vector3*)d;
00007FF95882087D  mov         rax,qword ptr [rbp+30h]  
00007FF958820881  vmovss      xmm0,dword ptr [rax+8]  
00007FF958820887  vmovsd      xmm1,qword ptr [rax]  
00007FF95882088C  vshufps     xmm1,xmm1,xmm0,44h  
00007FF958820892  vmovapd     xmmword ptr [rbp+20h],xmm1  
        }
````
__Method Test:__
````Assembly

var a = *(Numerics.Vector3*)(&value);
00007FF9588207F5  mov         al,byte ptr [rbp+38h]  
00007FF9588207F8  mov         byte ptr [rbp+20h],al  
        }
````

**Comment by mikedn** [comment]
Interesting, the code generated for method `Test` copies only one byte. That's not necessarily wrong because the variable `a` is not used anywhere. But it's a bit suspect, the compiler should either remove the entire copy or copy everything.

The current .NET Core build generates the following code:
```asm
       488D442418           lea      rax, bword ptr [rsp+18H]
       C4E17A104008         vmovss   xmm0, dword ptr [rax+8]
       C4E17B1008           vmovsd   xmm1, qword ptr [rax]
       C4E170C6C844         vshufps  xmm1, xmm0, 68
       C4E179290C24         vmovapd  xmmword ptr [rsp], xmm1
```
It's possible that the bug was already fixed but the fix is yet to reach .NET Framework.

@CarolEidt I'm pretty sure there was an issue that involved `Vector3` and this kind of casting but I can't find it. Maybe you remember it?

**Comment by CarolEidt** [comment]
> I'm pretty sure there was an issue that involved Vector3 and this kind of casting but I can't find it. Maybe you remember it?

No, but then I have a terrible memory for details.

**Comment by briansull** [comment]
I am not able to create a repro for this:
I am blocked on the first step.

1.  Install engenious via nuget or wrote your own packed Vector3 or use a other packed Vector3

**Comment by Gallimathias** [comment]
What's the problem? Can't you build your own packed Vector? @briansull

**Comment by briansull** [comment]
No I haven't ever done that

**Comment by jvbsl** [comment]
https://hastebin.com/ebegimomaf.cpp

apart from that this is actually a duplicate, as the other thread got moved over from roslyn repository to here
https://github.com/dotnet/coreclr/issues/16068

**Comment by briansull** [comment]
I still need the complete code with a simple compilable repro case.
 Ideally one that prints Pass or Fail would be great.

**Comment by jvbsl** [comment]
https://hastebin.com/kinixidemu.cs
printed on the debug console.
But you should know that its dependant on the used processor, for some processors it seems to work in both cases....
of course you would need the Numerics.Vectors nuget package as well:
https://www.nuget.org/packages/System.Numerics.Vectors

**Comment by briansull** [comment]
That still isn't a compilable repro.  (No Main and no using statements)
And if thefailure depends upon the exact hardware setup you have to mention that as well.

**Comment by jvbsl** [comment]
don't know what more you need except that doing it on your own would've been so much faster, but nvm. @Gallimathias is gonna make you a git repository.
Apart from that I need to admit, that I simply overread the "compilable" part and "repro" was for me "reproduction" and not repository or project^^
And it was already mentioned that it depends on the used hardware:
> Method TestB works on all systems no matter if 32 Bit or 64 Bit.
> Method Test does not work as 64 bit version on some systems. Could this be related to the CPU used?
and it's definitely not dependent on the hardware vendor, as it happens on both Intel and AMD.
I do have a i5-7200U e.g. and for me it works in both cases, @Gallimathias system is mentioned above as well and for him it doesn't work in both cases

**Comment by Gallimathias** [comment]
Attention repo is coming. Code and binary files are included. [SIMD Bug Check Tool](https://github.com/Gallimathias/SIMDBugTestTool) ;)

My CPU Intel(R) Core(TM) i7-5960X CPU @ 3.00GHz 

Test Results:
* Test 32 bit both is ok
* Test 32 Debug both is ok
* Test 64 Bit both failed
* Test 64 Debug direct casting failed

Test on AMD FX-6350 failed

Test Results for Intel(R) Xeon(R) CPU E5-2673 v4 2.30 GHz
* Test 32 bit both is ok
* Test 32 Debug both is ok
* Test 64 Bit both failed
* Test 64 Debug direct casting failed


Test Results for Intel(R) Core(R) i5-3337U CPU 1.80 GHz
* Test 32 bit both is ok
* Test 32 Debug both is ok
* Test 64 Bit both failed
* Test 64 Debug direct casting failed

I have more computers but I think the result will be the same. :D

**Comment by briansull** [comment]
Thanks for all your work on this issue .
We likely needed something like this.
I have cloned your repro and I have tried it out.

I don't see a failure using our current bits (will be 4.7.3) 
It does print failed on my older 4.7.1 (shipping) version:

 Image path: C:\WINDOWS\Microsoft.NET\Framework64\v4.0.30319\clrjit.dll
 ProductVersion:   4.7.2633.0

**Comment by briansull** [comment]
The fix for this issue requires both PR #15726 AND PR #15814

@CarolEidt

**Comment by RussKeldorph** [comment]
Fixed in coreclr.
